### PR TITLE
Expose a typescript & typescriptReact config

### PR DESCRIPTION
**Pull request by ismail-syed**
* This PR adds a consumable `typescript` & `typescriptReact` config for projects with `.ts` and `.tsx` files. 
* It's likely that some minor rules may have been missed, but I've disabled all the major one that conflicts with the ts parser. The parser is sometimes played nice and lists out which rule it failed on, other times it was all divide and conquer. 

Next steps: 
* Add `prettier`  plugins
* Update `lint` commands in `sewing-kit`. 
* Add `format` command in `sewing-kit`

Tophat instructions:

**NOTE:** The branches below are using beta releases of `eslint-plugin-shopify` with the work in this branch. 

---

### 🎩  in `sewing-kit`:
1. In `sewing-kit` checkout [`test-typescript-eslint-config`](https://github.com/Shopify/sewing-kit/commits/test-typescript-eslint-config) and run 
```bash
yarn install
```
2. `yarn run lint` to verify that it's linting as per the `typescript` config in this PR.

### 🎩 in `web`:

1. In `web` checkout [`test-typescript-eslint-config`](https://github.com/Shopify/web/commits/test-typescript-eslint-config) and run
```bash
yarn install
``` 

2. `yarn run lint` to verify that it's linting as per the `typescript` & `typescriptReact` config in this PR.

### CODE REVIEW

**Comment by ismail-syed** [comment]
In the [`test-typescript-eslint-config`](https://github.com/Shopify/web/commits/test-typescript-eslint-config) test branch in `web`, you'll notice that I added `*/**/*.d.ts` to the `.eslintignore`. That's because the parser had trouble parsing the typedef files. 

Going to investigate future. My guess is it's one of two things: 
* The typescript parser has trouble parsing typdef files. 
* There still may be a rule in our eslint config is conflicting with the parser.

File: lib/config/typescriptReact.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+var merge = require('merge');
+
+module.exports = {
+  extends: [
+    // react should come before typescript
```
Comment (ismail-syed) [inline lib/config/typescriptReact.js]: `plugin:shopify/typescript`'s config has the disabled rules that override the standard `esnext` rules that conflict with the parser. 
Listing `plugin:shopify/typescript` last, ensures that the disabled rules stay disabled, instead of being re-enabled by the react config.

File: lib/config/rules/typescript.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  // enforces the use of as Type assertions instead of <Type> assertions. (no-angle-bracket-type-assertion from TSLint)
+  'typescript/no-angle-bracket-type-assertion': 'error',
+
+  // disallows the use of custom TypeScript modules and namespaces
+  'typescript/no-namespace': 'off',
+
+  // disallows the use of variables before they are defined.
+  'typescript/no-use-before-define': 'off',
+
+  // enforces the use of the keyword namespace over module to declare custom TypeScript modules. (no-internal-module from TSLint) <FEEDBACK>
```
Comment (ismail-syed) [inline lib/config/rules/typescript.js]: Figure out why I labelled this with `<FEEDBACK>`, lol.

**Comment by GoodForOneFare** [comment]
Can you add an exception to [our `space-before-function-paren`](https://github.com/Shopify/eslint-plugin-shopify/blob/870ec7f263231923824f11ef924594e1e60ab9e3/lib/config/rules/stylistic-issues.js#L159) config, please?  There's an [`asyncArrow` option](https://eslint.org/docs/rules/space-before-function-paren#options) that'll allow us to keep:
`async () => {`.

**Comment by GoodForOneFare** [comment]
In `sewing-kit/src/tools/jest/index.ts`:
```export default async function runJest(
  workspace: Workspace,
  {
    watch = !workspace.env.isCI,
    debug = false,
    testRegex,
    ...options,
  }: Partial<Options> = {},
```

The `...options` line is flagged as `[eslint] Unexpected trailing comma. (comma-dangle)`.  A little annoying, but not a game changer.

**Comment by ismail-syed** [comment]
Added the `asynArrow` change as a separate PR since this PR is scoped around the ts configs :p https://github.com/Shopify/eslint-plugin-shopify/pull/43

**Comment by GoodForOneFare** [comment]
`web`'s `FailedRequirementsList.tsx` is autofixed to have a really funky layout:

```
    import * as React from 'react';

    import {autobind, compose} from 'utilities/decorators';
    import {withTranslation, WithTranslationProps} from '@shopify/i18n';

    import {
  Collapsible,
  Icon,
  Link,
} from '@shopify/polaris';

    import {SalesChannelAppsQueryData} from '../SalesChannelQuery.graphql';
    import * as styles from '../AddSalesChannelModal.scss';

    interface Props {
  failedRequirements: SalesChannelAppsQueryData.FailedRequirement[],
}
```

The entire file is all messed up.

**Comment by GoodForOneFare** [comment]
`packages/@shopify/polaris-next/components/DataTable/DataTable.tsx` ends up with this in it 
😕 :

```
export interface Props {
  headings: string[],
  rows: (string | number)[][],
  totals?: (
        />import * as React from 'react';
import {autobind} from '@shopify/javascript-utilities/decorators';
import {addEventListener, removeEventListener} from '@shopify/javascript-utilities/events';
import {classNames} from '@shopify/react-utilities/styles';

import DataTableCell from './DataTableCell';
import DataTableNavigation from './DataTableNavigation';
import * as styles from './DataTable.scss';
```

`Nav/User.tsx`, `Table/Table.tsx` end up with similar breaking changes.

File: lib/config/rules/typescript.js (hunk 2 above)
Comment (ismail-syed) [reply lib/config/rules/typescript.js]: This rule is [disabled](https://github.com/Shopify/tslint-config-shopify/blob/37055ada39ebd9d531ec61ca62159c6ccefb4f65/src/rules/typescript.ts#L26) in our tslint config. Adding `<FEEDBACK>` must have been an accidental paste.

File: lib/config/typescript.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  plugins: [
+    'typescript',
+  ],
+
+  parserOptions: {
+    ecmaVersion: 2017,
+    sourceType: 'module',
+  },
+
+
```
Comment (ismail-syed) [inline lib/config/typescript.js]: 🔥

**Comment by ismail-syed** [comment]
> In the [`test-typescript-eslint-config`](https://github.com/Shopify/web/commits/test-typescript-eslint-config) test branch in `web`, you'll notice that I added `*/**/*.d.ts` to the `.eslintignore`. That's because the parser had trouble parsing the typedef files. 

After digging a bit deeper, I found out that the parsing errors are coming from the [`babel-eslint`](https://github.com/Shopify/eslint-plugin-shopify/blob/master/lib/config/esnext.js#L5) parser that's specified in the `esnext` config. 

Follow up: In https://github.com/Shopify/web/blob/master/config/.eslintrc.js#L3, `plugin:shopify/esnext` config uses the `babel-eslint` parser.
Since some of the typdefs are coming from `/config`, eslint is taking the `/config/.eslintrc.js` config with higher precedence than what we specified in the `package.json`

**Review by GoodForOneFare — COMMENTED** [review]
So, definite blockers:
- Mangled code in `web`
- Parse errors in `admin` (mostly for explicit typecasts)

I think we either have to wait for the parser to improve, or improve it ourselves.  Let's get together with @lemonmade and discuss next steps.

File: lib/config/all.js [hunk 4]
```diff
@@ -32,6 +32,8 @@ module.exports = {
     'react',
     'shopify',
     'sort-class-members',
+    'typescript',
+    'typescriptReact',
```
Comment (lemonmade) [inline lib/config/all.js]: There's no way for these configurations to know about what other configs are on, and to have the `shopify/react` config update accordingly when it is used alongside `shopify/typescript`?

File: lib/config/rules/typescriptReact.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+  // Disable JS specific rules
+  'react/jsx-filename-extension': 'off',
+  'react/default-props-match-prop-types': 'off',
+
+  // Breaks typescript-eslint-parser
+  'react/jsx-indent': 'off',
```
Comment (lemonmade) [inline lib/config/rules/typescriptReact.js]: This is sad :(

File: lib/config/rules/typescriptReact.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+  'react/no-typos': 'off',
```
Comment (lemonmade) [inline lib/config/rules/typescriptReact.js]: Same here :(

File: lib/config/rules/typescript.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+  // enforces the use of the keyword namespace over module to declare custom TypeScript modules. (no-internal-module from TSLint)
+  'typescript/prefer-namespace-keyword': 'off',
+
+  // disallows the use of type aliases. (interface-over-type-literal from TSLint)
+  // breaks `export type Message = string | ((colorizer: any) => string);`
+  'typescript/no-type-alias': 'off',
+
+  // enforces a standard member declaration order. (member-ordering from TSLint) <FEEDBACK>
+  'typescript/member-ordering': 'error',
```
Comment (lemonmade) [inline lib/config/rules/typescript.js:36]: What is the ordering?

File: lib/config/rules/typescript.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+  // disallows parameter properties in class constructors. (no-parameter-properties from TSLint)
+  'typescript/no-parameter-properties': 'off',
+
+  // enforces PascalCased class and interface names. (class-name from TSLint)
+  'typescript/class-name-casing': 'error',
+
+  // enforces a member delimiter style in interfaces and type literals.
+  'typescript/member-delimiter-style': ['error', {
+    delimiter: 'comma',
```
Comment (lemonmade) [inline lib/config/rules/typescript.js]: I think we'll want `;` for Prettier, no?

File: lib/config/rules/typescript.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        delimiter: 'comma',
+      },
+    },
+  }],
+
+  // disallows the declaration of empty interfaces. (no-empty-interface from TSLint)
+  'typescript/no-empty-interface': 'off',
+
+
+  // Disable JS specific rules
```
Comment (lemonmade) [inline lib/config/rules/typescript.js]: These aren't all JS-specific, they are just things handled by TypeScript

File: lib/config/rules/typescript.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  'no-undef': 'off',
+  'import/no-unresolved': 'off',
+  'no-unused-expressions': 'off',
+  'no-unused-vars': 'off',
+  'no-useless-constructor': 'off',
+  'no-shadow': 'off',
+  'no-use-before-define': 'off',
+  'import/no-extraneous-dependencies': 'off',
+
+  // Breaks typescript-eslint-parser
```
Comment (lemonmade) [inline lib/config/rules/typescript.js]: Can we describe why each of these break it? Have we raised issues for them?

File: lib/config/typescriptReact.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    'plugin:shopify/react',
+    'plugin:shopify/typescript',
+  ],
+
+  parserOptions: {
```
Comment (lemonmade) [inline lib/config/typescriptReact.js]: Why are these needed?

File: lib/config/all.js [hunk 12]
```diff
@@ -55,6 +57,8 @@ module.exports = {
     require('./rules/promise'),
     require('./rules/react'),
     require('./rules/shopify'),
-    require('./rules/sort-class-members')
+    require('./rules/sort-class-members'),
+    require('./rules/typescript'),
+    require('./rules/typescriptReact')
```
Comment (lemonmade) [inline lib/config/all.js]: Looks like we do dash-case for these kinds of names

**Review by lemonmade — CHANGES_REQUESTED** [review]

File: lib/config/all.js (hunk 4 above)
Comment (ismail-syed) [reply lib/config/all.js]: I think that would only be possible if we are linting code through the eslint api, instead specifying configs.

File: lib/config/rules/typescript.js (hunk 8 above)
Comment (ismail-syed) [reply lib/config/rules/typescript.js]: Yah, I was planning to update this when we add the prettier config.

File: lib/config/typescriptReact.js (hunk 11 above)
Comment (ismail-syed) [reply lib/config/typescriptReact.js]: Oh yes, they aren't needed since the configs are extended.

File: lib/config/rules/typescript.js (hunk 10 above)
Comment (ismail-syed) [reply lib/config/rules/typescript.js]: Dissecting why each of these breaks with its respective line in our source can be fairly cumbersome. 

Eg. If I re-enable `array-callback-return`, and run the ts config linter in web, the following error shows up:
```
╰─$ yarn run lint
yarn run v1.0.1$ eslint . --ext .ts --ext .tsx --max-warnings 0
Cannot read property 'type' of null
TypeError: Cannot read property 'type' of null
    at EventEmitter.onCodePathStart (/Users/ismailsyed/src/github.com/Shopify/web/node_modules/eslint/lib/rules/array-callback-return.js:193:34)
    at emitTwo (events.js:130:20)
    at EventEmitter.emit (events.js:213:7)
    at processCodePathToEnter (/Users/ismailsyed/src/github.com/Shopify/web/node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:360:30)
    at CodePathAnalyzer.enterNode (/Users/ismailsyed/src/github.com/Shopify/web/node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:603:9)
    at nodeQueue.forEach.traversalInfo (/Users/ismailsyed/src/github.com/Shopify/web/node_modules/eslint/lib/linter.js:985:32)
    at Array.forEach (<anonymous>)
    at Linter._verifyWithoutProcessors (/Users/ismailsyed/src/github.com/Shopify/web/node_modules/eslint/lib/linter.js:981:19)
    at preprocess.map.textBlock (/Users/ismailsyed/src/github.com/Shopify/web/node_modules/eslint/lib/linter.js:1021:35)
    at Array.map (<anonymous>)
error Command failed with exit code 1.
```

I've been using trial & error to find out which line breaks it. I can do this for every breaking rule if time permits.

Comment (ismail-syed) [reply lib/config/rules/typescript.js]: https://github.com/eslint/typescript-eslint-parser/issues/389

**Comment by ismail-syed** [comment]
* Formatting issues from `FailedRequirementsList.tsx` are coming from the `react/jsx-closing-tag-location` fixer.
* Formatting issues from `DataTable.tsx` are coming from the `react/jsx-wrap-multilines` fixer. 

Same with `Nav/User.tsx`, `Table/Table.tsx`

Looks like we're getting hit by https://github.com/eslint/typescript-eslint-parser/issues/343

**Comment by ismail-syed** [comment]
> Parse errors in admin (mostly for explicit typecasts)

The parse errors were due to nested `eslintrc.js` files not extending the `typescript` and `typescript-react` configs. Changes [here](https://github.com/Shopify/shopify/commit/ffe9846332f6fe6867de1da725247177aa6bec32) should not show parse errors. These `eslintrc.js` files are kind of wonky in the sense that they try to be a js config and a ts config. In practice, I think we may have to add a separate config for ts.

**Comment by ismail-syed** [comment]
Closing this since it's WIP atm.
